### ISSUE TITLE
Subtraction of extruded mesh not working

### ISSUE THREAD

**Issue by kenianbei** [samalexander/three-csg-ts#37]
**Describe the bug**
Subtraction of extruded geometries isn't working correctly in three v0.131.1 and three-csg-ts v3.1.6. However if downgrading to  three v0.129.0 and three-csg-ts v3.1.4 it works normally. What changed?

**To Reproduce**
Demo (not working version): https://codesandbox.io/s/extruded-subtraction-demo-opss0?file=/src/index.ts
Demo (working version): https://codesandbox.io/s/extruded-subtraction-demo-w-old-version-1mw34?file=/src/index.ts:1296-1326

**Expected behavior**
Extruded geometries should be subtracted normally.

**Comment by kenianbei** [comment]
My guess from experience with openjscad is that the face normals are not correct for some reason with extrusion? Any tips on fixing this?

**Comment by samalexander** [comment]
This issue seems to be introduced by v3.1.6, as your demo works with v3.1.5 - see [here](https://codesandbox.io/s/extruded-subtraction-demo-forked-vs9qz). I'll take another look at the changes introduced by v3.1.6 and test against your code. In the meantime, you can just stick to v3.1.5.

**Comment by lichr** [comment]
Hi There,  I ran into the same problem and did some investigation on it. 

I believe a change on v3.1.6 caused the issue, as you mentioned. It is possible that the issue is not specific to extruded geometries, it could be a problem how geometries with groups are handled. 

I worked on a fix and the test result seems good, I created a PR please take a look, thanks.

**Comment by samalexander** [comment]
Fixed in v3.1.8 - thanks @lichr for the PR!

**Comment by kenianbei** [comment]
This fixed the box extrude demo, thank you! However I found another instance of this behavior, as shown here:  https://codesandbox.io/s/cylinder-subtraction-7lfvg?file=/src/index.ts

It seems like when the cylinder is positioned some distance from the center it causes the same behavior. Just found this out after testing the new version.

**Comment by manthrax** [comment]
@kenianbei 

can you replace the line:


     sub.position.set( ...position );
  
 with:
       sub.position.set(10,0,0);
       
and tell me what you see?

This might be a typescript problem?

**Comment by manthrax** [comment]
I don't understand what exactly is happening there...

**Comment by lichr** [comment]
Haven't drilled too deep into the BSP implementation, looks like rotating the sphere for a little bit like this can fix the problem:

```js
  shape.ellipse(
    0,
    0,
    width / 2,
    height / 2,
    degToRad(0),
    degToRad(360),
    true,
    degToRad(0.001) // dirty fix here
  );

```

**Comment by lichr** [comment]
In @kenianbei 's example the extruded geometry has overlapped vertexes like this:
![image](https://user-images.githubusercontent.com/19759171/132141884-69e980ae-3c4c-41c7-8e87-c30f6c95aeeb.png)

This results in a polygon that is so narrow that degrades into a line, so that its normal can not be determined. I created a PR to filter out such polygons before building the BSP tree, looks like this fixes the problem.

**Comment by samalexander** [comment]
Fixed in v3.1.9 - thanks again @lichr for the PR!
